src/nomad_simulations/properties/band_gap.py

```diff
@@ -0,0 +1,147 @@
+#
+# Copyright The NOMAD Authors.
+#
+# This file is part of NOMAD. See https://nomad-lab.eu for further info.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import numpy as np
+from structlog.stdlib import BoundLogger
+import pint
+from typing import Optional
+
+from nomad.units import ureg
+from nomad.metainfo import Quantity, MEnum, Section, Context
+
+from ..physical_property import PhysicalProperty
+
+
+class ElectronicBandGap(PhysicalProperty):
+    """
+    Energy difference between the highest occupied electronic state and the lowest unoccupied electronic state.
+    """
+
+    iri = 'https://fairmat-nfdi.github.io/fairmat-taxonomy/#/ElectronicBandGap'
+
+    # ? can `type` change character depending on the `variables`?
+    type = Quantity(
+        type=MEnum('direct', 'indirect'),
+        description="""
+        Type categorization of the `ElectronicBandGap`. This quantity is directly related with `momentum_transfer` as by
+        definition, the electronic band gap is `'direct'` for zero momentum transfer (or if `momentum_transfer` is `None`) and `'indirect'`
+        for finite momentum transfer.
+
+        Note: in the case of finite `variables`, this quantity refers to all of the `value` in the array.
+        """,
+    )
+
+    momentum_transfer = Quantity(
+        type=np.float64,
+        shape=[2, 3],
+        description="""
+        If the `ElectronicBandGap` is `'indirect'`, the reciprocal momentum transfer for which the band gap is defined
+        in units of the `reciprocal_lattice_vectors`. The initial and final momentum 3D vectors are given in the first
+        and second element. Example, the momentum transfer in bulk Si2 happens between the Γ and the (approximately)
+        X points in the Brillouin zone; thus:
+            `momentum_transfer = [[0, 0, 0], [0.5, 0.5, 0]]`.
+
+        Note: this quantity only refers to scalar `value`, not to arrays of `value`.
+        """,
+    )
+
+    spin_channel = Quantity(
+        type=np.int32,
+        description="""
+        Spin channel of the corresponding `ElectronicBandGap`. It can take values of 0 or 1.
+        """,
+    )
```

Comment:
I would prefer not. It is totally fine to have categories, but really we don't gain anything from having `spin_channel` as the unique quantity under some over-nested tree. Furthermore, there is no normalization done in different spin channels.

Let's simplify and keep it flat.

Reply:
> we don't gain anything from having spin_channel as the unique quantity under some over-nested tree

Uhm, that's not quite what I meant. Rather something long the lines of the following (contrived) example:

```python
output = [
    ElectronicStructure(k_mesh=k_grid_1, spin_channel=[0], dos=DOS(...), band_gapElectronicBandGap(...)),
    ElectronicStructure(k_mesh=k_grid_1, spin_channel=[1], dos=DOS(...), band_gapElectronicBandGap(...)),
    ElectronicStructure(k_mesh=k_path_1, spin_channel=[0, 1], band_structure=BandStructure(...), dos=DOS(...), band_gapElectronicBandGap(...)),
]
```

You see? This setup emphasizes band structures, DOS, band gaps that come from the same setup.
If we flatten everything out, I first have to scan all `output` to find the matching `BandStructure` and `DOS`.
I'm really thinking of the user here, since I don't want them to have to write `for`-loops, just to extract this kind of information. This also predefines the packages for visualization.

Reply:
I am not sure what is that you are writing.

A real case example would look like:
```python
simulation = Simulation(
    model_system=...,
    model_method=...,
    outputs=Outputs(
        electronic_band_gap=[
            ElectronicBandGap(spin_channel=0, physical_property_ref=electronic_dos[0], value=...),
            ElectronicBandGap(spin_channel=1, physical_property_ref=electronic_dos[1], value=...),
            ElectronicBandGap(physical_property_ref=electronic_band_structure[0], value=...),
        ],
        electronic_dos=[
            ElectronicDOS(spin_channel=0, value=...),
            ElectronicDOS(spin_channel=1, value=...),
            ElectronicDOS(physical_property_ref=electronic_band_structure[0], value=...),
        ],
        electronic_band_structure=[
            ElectronicBandStructure(value=...),
        ]
    )
)
```

But I doubt this example is realistic. In any case, you will not get rid of writing for loops in your example (if I understand correctly, you want 3 sections for electronic structure under outputs). Furthermore, why to do this classification for the electronic properties and not for other categories? What about when a property is shared amongst 2 or more categories?

I am trying to avoid that.

Reply:
Okay, but in the example that you lay out, how does the data scientist know

1. the spin channel of `ElectronicBandGap(physical_property_ref=electronic_band_structure[0], value=...)`?
Should she read its reference? Should it store the `spin_channel` of the property its refering to?
2. if the material (not the calculation) is spin-polarized or not? Is it now up to the data scientist to write a check herself?

Reply:
Okay, I saw that you more or less addressed this in your comment below + commits.